oauth2client/contrib/flask_util.py

```diff
@@ -443,7 +443,12 @@ def credentials(self):
 
     def has_credentials(self):
         """Returns True if there are valid credentials for the current user."""
-        return self.credentials and not self.credentials.invalid
```

Comment:
Agreed, also surprisingly, invalid is not even set before a failed request even if there is no refresh token present. At the every least, the client can detect if the access token is expired, and there isn't a refresh token, and set invalid preemptively.